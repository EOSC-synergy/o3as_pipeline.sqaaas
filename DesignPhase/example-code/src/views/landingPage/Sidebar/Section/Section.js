import React, { useState } from "react";
import ListBox from '../inputs/ListBox';
import TextBox from '../inputs/TextBox';
import defaultStructure from '../../../../config/defaultConfig.json';

function mapNameToComponent(name) {
    return 0;
}

export default function Section(props) {
    
    const [expanded, setExpanded] = useState(true);
    props.components;

    return <div>
        <ListBox text="Plot Type" items={["T", "O", "D", "O"]} selected={plotType} />
        <TextBox text="Plot Name" onChange={handlePlotNameChange}/>
        {props.components.array.forEach(element => {
            mapNameToComponent(element);
        })}

        {defaultStructure.forEach(c => <>
            {/*Components from defaultStructure*/}
        </>)}
    </div>;
}