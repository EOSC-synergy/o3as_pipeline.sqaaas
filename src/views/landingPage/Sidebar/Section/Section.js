import React, { useState } from "react";
import defaultStructure from '../../../../config/defaultConfig.json';
import LocationSelector from "../InputComponents/LatitudeBandSelector/LatitudeBandSelector";
import ModelGroupConfigurator from "../InputComponents/ModelGroupConfigurator/ModelGroupConfigurator";
import OffsetConfigurator from "../InputComponents/OffsetConfigurator/OffsetConfigurator";
import PlotNameField from "../InputComponents/PlotNameField/PlotNameField";
import ReferenceYearSlider from "../InputComponents/ReferenceYearSlider/ReferenceYearSlider";
import RegionSelector from "../InputComponents/RegionSelector/RegionSelector";
import TimeCheckBoxGroup from "../InputComponents/TimeCheckboxGroup/TimeCheckboxGroup";
import XAxisSlider from "../InputComponents/XAxisSlider/XAxisSlider";
import YAxisSlider from "../InputComponents/YAxisSlider/YAxisSlider";
import PropTypes from 'prop-types'; 
import LatitudeBandSelector from "../InputComponents/LatitudeBandSelector/LatitudeBandSelector";
import ReferenceModelSelector from "../InputComponents/ReferenceModelSelector/ReferenceModelSelector";
import { Accordion } from '@mui/material';
import AccordionSummary from '@mui/material/AccordionSummary';
import AccordionDetails from '@mui/material/AccordionDetails';
import Typography from '@mui/material/Typography';
import ExpandMoreIcon from '@mui/icons-material/ExpandMore';

/**
 * an expandable section containing a list of inputComponents as well as a name
 * @param {Object} props
 * @param {Array of String} props.components -> an array containing a string representation of all components that should be plotted
 * @param {String} props.name -> the name of the section
 * @param {function} props.reportError -> used for error handling
 * @param {boolean} props.isExpanded -> whether this section should be expanded
 * @param {function} props.onCollapse -> function to collapse this section
 * @param {function} props.onExpand -> function to expand this section
 * @returns {JSX} an accordeon that once expanded displays the components specified by the config files and the API doc
 */
function Section(props) {

    /**
     * maps a given name to a corresponding component from the ./InputComponents folder
     * if there is no component with the given name, calls the props.reportError function
     * @public
     * @param {String} name the name of the component
     * @param {int} key a unique key for the given input component
     * @returns a component from the './InputComponents
     */
    function mapNameToComponent(name, key) {
        switch (name){
            case "ModelGroupConfigurator":
                return <ModelGroupConfigurator key={key} />;
            case "OffsetConfigurator":
                return <OffsetConfigurator key={key}  />;
            case "PlotNameField":
                return <PlotNameField key={key}  />;
            case "ReferenceYearSlider":
                return <ReferenceYearSlider key={key}  />;
            case "TimeCheckBoxGroup":
                return <TimeCheckBoxGroup key={key}  />;
            case "XAxisSlider":
                return <XAxisSlider key={key}  />;
            case "YAxisSlider":
                return <YAxisSlider key={key}  />;
            case "RegionSelector":
                return <RegionSelector key={key}  />;
            case "LatitudeBandSelector":
                return <LatitudeBandSelector key={key}  />;
            case "ReferenceModelSelector":
                return <ReferenceModelSelector key={key} />;
            default:
                if (props.reportError) {
                    props.reportError(`Section ${props.name} found no match for an input component ${name}`);
                }
        } 
    }

    if (!props.components) {
       if(props.reportError) {
           props.reportError(`Section ${props.name} was provided with no components`)
        }
        return <></>;
    }

    return (
    <Accordion data-testid="section">
        <AccordionSummary
          expandIcon={<ExpandMoreIcon />}
          aria-controls="panel1a-content"
          id="panel1a-header"
        >
            {props.name}
        </AccordionSummary>
        <AccordionDetails>
            <>
                {props.components.map((element, idx) => {
                    return (
                        <>
                            {mapNameToComponent(element, idx)}
                            <br />
                        </>
                    )
                })}
            </>
        </AccordionDetails>
    </Accordion>);
}

Section.propTypes = {
    name: PropTypes.string,
}

export default Section;