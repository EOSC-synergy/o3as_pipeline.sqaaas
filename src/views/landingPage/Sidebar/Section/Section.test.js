import React from 'react';
import ReactDOM from 'react-dom';
import Section from './Section';
import renderer from 'react-test-renderer';
import {render, screen} from '@testing-library/react';
import defaultStructure from '../../../../config/defaultConfig.json';
import "@testing-library/jest-dom/extend-expect";

it('Section renders without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(<Section />, div);
});

const testPropTypes = (component, propName, arraysOfTestValues, otherProps) => {
    console.error = jest.fn();
    const _test = (testValues, expectError) => {
        for (let propValue of testValues) {
            console.error.mockClear();
            React.createElement(component, {...otherProps, [propName]: propValue});
            expect(console.error).toHaveBeenCalledTimes(expectError ? 1 : 0);
        }
    };
    _test(arraysOfTestValues[0], false);
    _test(arraysOfTestValues[1], true);
};

// report error if no components are entered
it('Section reports error if no components are passed to render', () => {
    const reportError = jest.fn();
    const name = "1";
    const div = document.createElement('div');
    ReactDOM.render(<Section reportError={reportError} name={name} isExpanded={false}/>, div);
    expect(reportError).toHaveBeenCalledWith(`Section ${name} was provided with no components`)
})

// report error if invalid components are entered
it('Section reports error if invalid component is entered', () => {
    const reportError = jest.fn();
    const name = "1";
    const compName = "blob";
    const div = document.createElement('div');
    ReactDOM.render(<Section reportError={reportError} components={[compName]} name={name} isExpanded={false}/>, div);
    expect(reportError).toHaveBeenCalledWith(`Section ${name} found no match for an input component ${compName}`)
})

// renders props.name
it('Section renders the given name', () => {
    const name = "Test";
    render(<Section components={["LatitudeBandSelector"]} name={name} isExpanded={false}/>);
    expect(screen.getByTestId('section')).toHaveTextContent(name.toUpperCase());
})


// Snapshot test
it('Section renders correctly from config file', () => {
    for (let section in defaultStructure.sections) {
        const tree = renderer
            .create(<Section name={section.name} reportError={() => {}} components={section.components} isExpanded={false}/>)
            .toJSON();
        expect(tree).toMatchSnapshot();
    }
});