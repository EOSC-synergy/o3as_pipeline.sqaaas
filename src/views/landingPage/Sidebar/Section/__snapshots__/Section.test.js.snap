// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Section renders correctly from config file 1`] = `null`;

exports[`Section renders correctly from config file 2`] = `null`;

exports[`Section renders correctly from config file 3`] = `null`;

exports[`Section renders correctly from config file 4`] = `null`;
